File: analysis/import/CSVImporter/src/main/java/de/maibornwolff/codecharta/importer/csv/CSVHeader.java
package de.maibornwolff.codecharta.importer.csv;

import java.util.HashMap;
import java.util.Map;
import java.util.Set;

public class CSVHeader {
    public static final String PATH_COLUMN_NAME = "path";
    private final Map<Integer, String> headerMap;

    public CSVHeader(String[] header) {
        headerMap = new HashMap<>();
        for (int i = 0; i < header.length; i++) {
            if (header[i] == null || header[i].isEmpty()) {
                System.err.println("Ignoring column number " + i + " (counting from 0) as it has no column name.");
            } else if (headerMap.containsValue(header[i])) {
                System.err.println("Ignoring column number " + i + " (counting from 0) with column name " + header[i] + " as it duplicates a previous column.");
            } else {
                headerMap.put(i, header[i]);
            }
        }

        if (headerMap.isEmpty()) {
            throw new IllegalArgumentException("Header is empty.");
        }
    }

    public Set<Integer> getColumnNumbers() {
        return headerMap.keySet();
    }

    public String getColumnName(int i) {
        String columnName = headerMap.get(i);
        if (columnName == null) {
            throw new IllegalArgumentException("No " + i + "th column present.");
        }
        return columnName;
    }

    public int getPathColumn() {
        return headerMap.keySet().stream()
                .filter(i -> headerMap.get(i).equalsIgnoreCase(PATH_COLUMN_NAME))
                .findFirst()
                .orElse(headerMap.keySet().stream().findFirst().get());
    }
}

File: analysis/import/CSVImporter/src/main/kotlin/de/maibornwolff/codecharta/importer/csv/CSVHeader.kt
package de.maibornwolff.codecharta.importer.csv

import mu.KotlinLogging
import java.util.*

class CSVHeader(private val header: Array<String?>, private val pathColumnName: String = "path") {
    private val logger = KotlinLogging.logger {}
    private val headerMap: MutableMap<Int, String>
    val columnNumbers: Set<Int>
        get() = headerMap.keys
    val pathColumn: Int
        get() = headerMap.keys.firstOrNull { i -> headerMap[i].equals(pathColumnName, ignoreCase = true) }
            ?: headerMap.keys.first()

    init {
        headerMap = HashMap()
        for (i in header.indices) {
            when {
                header[i].isNullOrEmpty() ->
                    logger.warn { "Ignoring ${i + 1}-th column number due to: Column has no name." }
                headerMap.containsValue(header[i]) ->
                    logger.warn { "Ignoring ${i + 1}-th column number due to: Column name '${header[i]}' duplicates a previous column." }
                else ->
                    headerMap[i] = header[i]!!
            }
        }

        if (headerMap.isEmpty()) {
            throw IllegalArgumentException("Header is empty.")
        }
    }

    fun getColumnName(i: Int): String {
        return headerMap[i] ?: throw IllegalArgumentException("No ${i + 1}-th column present.")
    }
}
